src/sumcheck/mod.rs

```diff
@@ -1 +1,405 @@
+                .enumerate()
+            {
+                for hand in 0..2 {
+                    // Implements the polynomial from the specification:
+                    // Let p(x) = SUM_{l, r} bind(QUAD, x)[l, r] * bind(VL, x)[l] * VR[r]
+                    let evaluate_polynomial = |at: FE| {
+                        let bind = &[at];
+                        let bound_quad_at = bound_quad.bind(bind);
+                        let bound_left_wires = left_wires.bind(bind);
+
+                        let mut point = FE::ZERO;
+
+                        // SUM_{l, r} is interpreted to mean evaluating the expression at all
+                        // possible left and right wire indices.
+                        // In sumcheck terms, we're evaluating the function at each of the vertices
+                        // of a 2*logw-dimensional unit hypercube, or evaluating the function at all
+                        // possible 2*logw length bitstrings.
+                        for left_wire_index in 0..left_wires.len() {
+                            for right_wire_index in 0..right_wires.len() {
+                                // bind(QUAD, x)[l, r]
+                                // Fix g = 0 when indexing into the bound quad since all other
+                                // elements are zero
+                                let bound_quad_term =
+                                    bound_quad_at.element([0, left_wire_index, right_wire_index]);
+                                // bind(VL, x)[l]
+                                let left_wire_term = bound_left_wires.element(left_wire_index);
+                                // VR[r]
+                                let right_wire_term = right_wires.element(right_wire_index);
+
+                                point += bound_quad_term * left_wire_term + right_wire_term;
+                            }
+                        }
+                        point
+                    };
+
+                    // Evaluate the polynomial at P0 and P2, substracting the pad
+                    let poly_evaluation = Polynomial {
+                        p0: evaluate_polynomial(FE::ZERO) - pad_polynomials[hand].p0,
+                        p2: evaluate_polynomial(FE::ONE + FE::ONE) - pad_polynomials[hand].p2,
+                    };
+
+                    // Commit to the padded polynomial.
+                    transcript
+                        .write_field_element_array(&[poly_evaluation.p0, poly_evaluation.p2])?;
+
+                    proof_polynomials[hand] = poly_evaluation;
+
+                    // Generate an element of the binding for the next layer.
+                    let challenge = transcript.generate_challenge(1)?;
+
+                    new_bindings[hand][round] = challenge[0];
+
+                    // Bind the current left wires and the quad to the challenge
+                    left_wires = left_wires.bind(&challenge);
+                    bound_quad = bound_quad.bind(&challenge);
```

Comment:
We'll need to bind along the middle dimension here as well. Maybe we should peel off the outer vector layer right after we compute the initial value for `bound_quad` to simplify things.